core/coreapi/dht.go

```diff
@@ -0,0 +1,241 @@
+	settings, err := caopts.DhtFindProvidersOptions(opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	dht, ok := api.node.Routing.(*ipdht.IpfsDHT)
+	if !ok {
+		return nil, ErrNotDHT
+	}
+
+	p, err = api.core().ResolvePath(ctx, p)
+	if err != nil {
+		return nil, err
+	}
+
+	c := p.Cid()
+
+	numProviders := settings.NumProviders
+	if numProviders < 1 {
+		return nil, fmt.Errorf("number of providers must be greater than 0")
+	}
+
+	outChan := make(chan peer.ID)
+	events := make(chan *notif.QueryEvent)
+	ctx = notif.RegisterForQueryEvents(ctx, events)
+
+	pchan := dht.FindProvidersAsync(ctx, c, numProviders)
+	go func() {
+		defer close(outChan)
+
+		sendProviders := func(responses []*pstore.PeerInfo) error {
+			for _, response := range responses {
+				select {
+				case outChan <- response.ID:
+				case <-ctx.Done():
+					return ctx.Err()
+				}
+			}
+			return nil
+		}
+
+		for event := range events {
+			if event.Type == notif.Provider {
+				err := sendProviders(event.Responses)
+				if err != nil {
+					return
+				}
+			}
+		}
+	}()
+
+	go func() {
+		defer close(events)
+		for p := range pchan {
+			np := p
+			notif.PublishQueryEvent(ctx, &notif.QueryEvent{
+				Type:      notif.Provider,
+				Responses: []*pstore.PeerInfo{&np},
+			})
+		}
+	}()
+
+	return outChan, nil
+}
+
+func (api *DhtAPI) Provide(ctx context.Context, path coreiface.Path, opts ...caopts.DhtProvideOption) error {
+	settings, err := caopts.DhtProvideOptions(opts...)
+	if err != nil {
+		return err
+	}
+
+	if api.node.Routing == nil {
+		return errors.New("cannot provide in offline mode")
+	}
+
+	if len(api.node.PeerHost.Network().Conns()) == 0 {
+		return errors.New("cannot provide, no connected peers")
+	}
+
+	c := path.Cid()
+
+	has, err := api.node.Blockstore.Has(c)
+	if err != nil {
+		return err
+	}
+
+	if !has {
+		return fmt.Errorf("block %s not found locally, cannot provide", c)
+	}
+
+	//TODO: either remove or use
+	//outChan := make(chan interface{})
+
+	//events := make(chan *notif.QueryEvent)
+	//ctx = notif.RegisterForQueryEvents(ctx, events)
+
+	/*go func() {
+		defer close(outChan)
+		for range events {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+		}
+	}()*/
+
+	//defer close(events)
+	if settings.Recursive {
+		err = provideKeysRec(ctx, api.node.Routing, api.node.DAG, []*cid.Cid{c})
+	} else {
+		err = provideKeys(ctx, api.node.Routing, []*cid.Cid{c})
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func provideKeys(ctx context.Context, r routing.IpfsRouting, cids []*cid.Cid) error {
+	for _, c := range cids {
+		err := r.Provide(ctx, c, true)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func provideKeysRec(ctx context.Context, r routing.IpfsRouting, dserv ipld.DAGService, cids []*cid.Cid) error {
+	provided := cid.NewSet()
+	for _, c := range cids {
+		kset := cid.NewSet()
+
+		err := dag.EnumerateChildrenAsync(ctx, dag.GetLinksDirect(dserv), c, kset.Visit)
```

Comment:
Same, could we get a "TODO: provide and enumerate in parallel with a small buffer".

Reply:
Will get that with https://github.com/ipfs/go-ipfs/pull/4333 for free